texar/torch/modules/tokenizers/pretrained_xlnet_tokenizer.py

```diff
@@ -0,0 +1,228 @@
+        self.__dict__ = d
+        self.sp_model = spm.SentencePieceProcessor()
+        self.sp_model.Load(self.vocab_file)
+
+    def preprocess_text(self, inputs: str) -> str:
+        if self.remove_space:
+            outputs = ' '.join(inputs.strip().split())
+        else:
+            outputs = inputs
+        outputs = outputs.replace("``", '"').replace("''", '"')
+
+        if not self.keep_accents:
+            outputs = unicodedata.normalize('NFKD', outputs)
+            outputs = ''.join([c for c in outputs if not
+            unicodedata.combining(c)])
+        if self.do_lower_case:
+            outputs = outputs.lower()
+
+        return outputs
+
+    def _tokenize(self, text: str,  # type: ignore
+                  sample: bool = False) -> List[str]:
+        text = self.preprocess_text(text)
+        if not sample:
+            pieces = self.sp_model.EncodeAsPieces(text)
+        else:
+            pieces = self.sp_model.SampleEncodeAsPieces(text, 64, 0.1)
+
+        new_pieces: List[str] = []
+        for piece in pieces:
+            if len(piece) > 1 and piece[-1] == ',' and piece[-2].isdigit():
+                cur_pieces = self.sp_model.EncodeAsPieces(
+                    piece[:-1].replace(SPIECE_UNDERLINE, ''))
+                if piece[0] != SPIECE_UNDERLINE and \
+                        cur_pieces[0][0] == SPIECE_UNDERLINE:
+                    if len(cur_pieces[0]) == 1:
+                        cur_pieces = cur_pieces[1:]
+                    else:
+                        cur_pieces[0] = cur_pieces[0][1:]
+                cur_pieces.append(piece[-1])
+                new_pieces.extend(cur_pieces)
+            else:
+                new_pieces.append(piece)
+
+        return new_pieces
+
+    def save_vocabulary(self, save_directory: str) -> Tuple[str]:
+        r"""Save the sentencepiece vocabulary (copy original file) to
+        a directory.
+        """
+        if not os.path.isdir(save_directory):
+            raise ValueError("Vocabulary path ({}) should be a "
+                             "directory".format(save_directory))
+        out_vocab_file = os.path.join(save_directory,
+                                      self._VOCAB_FILE_NAMES['vocab_file'])
+
+        if os.path.abspath(self.vocab_file) != os.path.abspath(out_vocab_file):
+            copyfile(self.vocab_file, out_vocab_file)
+
+        return (out_vocab_file,)
+
+    @property
+    def vocab_size(self) -> int:
+        return len(self.sp_model)
+
+    def _convert_token_to_id(self, token: str) -> int:
+        return self.sp_model.PieceToId(token)
+
+    def _convert_id_to_token(self, index: int) -> str:
+        token = self.sp_model.IdToPiece(index)
+        return token
+
+    def convert_tokens_to_string(self, tokens: List[str]) -> str:
+        out_string = ''.join(tokens).replace(SPIECE_UNDERLINE, ' ').strip()
+        return out_string
+
+    @staticmethod
+    def default_hparams() -> Dict[str, Any]:
+        return {
```

Comment:
I think we need docstring explaining each entry in the hparams dictionary, like we do for every other class.